book/src/programs.md

```diff
@@ -1,32 +1,67 @@
-program. The runtime also ensures accounts aren't written to if not owned
-by the client or program. Any writes to credit-only accounts are discarded
-unless the write was to credit tokens. Any user may credit other accounts
-tokens, regardless of account permission.
+As shown in the diagram above a client creates a program and compiles it to an
+ELF shared object containing BPF bytecode and sends it to the Solana cluster.
+The cluster stores the program locally and makes it available to clients via a
+*program ID*. The program ID is a *public key* generated by the client and is
+used to reference the program in subsequent transactions.
+
+## Interacting with On-chain Programs
+
+After a program has been deployed, a client utilizes it by sending transactions
+to the Solana cluster. A transaction contains one or more *instructions*. Each
+instruction references the program ID that corresponds to the program that the
+runtime should use to execute it. Instructions are executed atomically. If any
+instruction is invalid, any changes made within the transaction are discarded.
+
+## Storing State between Transactions
+
+If the program needs to store state between transactions, it does so using
+*accounts*. Accounts are similar to files in operating systems such as Linux.
+Like files, an account may hold arbitrary data and that data persists beyond
+the lifetime of a program. Also like files, an account includes metadata that
+tells the runtime who is allowed to access the data and how. Unlike files, the
+account includes metadata for the lifetime of the file. That lifetime is
+expressed in "tokens", which is a number of fractional native tokens, called
+*lamports*. Accounts are held in validator memory and pay "rent" to stay there.
+Each fullnode periodically scan all accounts and collects rent. Any account
+that drops to zero lamports is purged.
+
+If an account is marked "executable", it will only be used by a *loader* to run
+programs. For example, programs compiled to BPF are marked executable and
+loaded by the BPF loader. No program is allowed to modify the contents of an
+executable account.
+
+Accounts also include "owner" metadata. The owner is a program ID. The runtime
+grants the program write access to the account if its ID matches the owner. If
+an account is not owned by a program, the program is permitted to read its data
+and credit the account.
+
+In the same way that a Linux user uses a path to look up a file, a Solana
+client uses public keys to look up accounts. To create an account, the client
+generates a *keypair* and registers its public key using the CreateAccount
+instruction. Once registered, transactions reference account keys to grant
+programs access to accounts. The runtime grants programs read access by
+default. To grant write access, the client must either assign the account
+to a program ID or sign the transaction using the keypair's *secret key*. Since
```

Comment:
I'm going to change that to "assign the account to a program" to focus on the "what" instead of the "how". Setting the owner field to a program ID is how assigning an account achieved. I'm not sure if the "how" part belongs here or in a separate section. I think I'll leave it out for now.